### ISSUE TITLE
Sync between presenter mode and play mode is rough

### ISSUE THREAD

**Issue by cyyynthia** [slidevjs/slidev#1987]
**Is your feature request related to a problem? Please describe.**
I've been playing with slidev for a bit, and one thing I always get bitten by is how limited the sync between presenter mode and play mode can be.

Here are a few examples I encountered, in no particular order:
- TwoSlash doesn't show up in presenter mode at all, and hovering elements don't trigger TwoSlash on the play view
- Stateful component interactions are not shared (e.g. the Counter demo example)
- Draggable elements transformations are not shared
  - This was quite annoying for me, as I tried to make a slide with such kind of elements to make a "manual" interactive demo. It seems that's not the intended usage of v-drag, but I think it'd be a cool QoL addition (without persisting positions, too)
- Monaco editor (and run results) are not synced
  - This is also a bit annoying when trying to do something interactive, and have whatever you're showcasing be visible "live" on screen

I also ran into some edge cases when testing locally wrt syncing drawings, especially when drawing sometimes on the play view and sometimes on the presenter view. It happened only once, so I'm unsure if it's worth investigating based on my sole experience and it might as well be a fluke.

*For the record, I peeked at the code if there was E2E tests, but since they're Cypress and it is \*not so great\*, it's not quite straightforward to set real multiplayer E2E tests up :(*

**Describe the solution you'd like**
While I'm fully aware replicating 1:1 mouse interactions is not (easily) feasible, having basic interactions such as mouseenter/mouseleave and click from a presenter view be reflected on the play views would be a plus. It may be useful to some to limit mouse event mirroring when the cursor is enabled.

Support for syncing draggable elements would be super cool for "playing" with elements on a slide within presenter view and have the play view reflect those changes in real time.

**Describe alternatives you've considered**
Point and move things on the play view instead. It certainly works, but is nowhere near as comfortable as just interacting with the computer in front of you when the play view is projected behind you.

**Comment by antfu** [comment]
Yeah we are aware of that. But I don't have a good solution for that in mind yet. Any suggestions? 

I don't know if it's possible to simulate the cursor events. Or maybe we could use screen-capturing to streaming back the user view?

**Comment by cyyynthia** [comment]
I've had some success sending fake events via dispatchEvent to make TwoSlash work with the "proxied" cursor (see attached patch), I assume clicks can be forwarded in a similar fashion. It won't cut it for css `:hover` states without extra work, but that can be a relatively simple way to make interactions shared without much work; it'd probably be best to be aware of whether the events have been preventDefault'd or not, too.

For draggable elments and Monaco, perhaps createSyncState can be the "easy" solution to make state shared across opened views but I haven't been wandering in the source code for long enough to know whether it's "as easy as that" or if there's more to it.

**Comment by antfu** [comment]
I am wondering if we should go the other way around, to have users' cursor on focus on the slides, where the presenter is more like a viewer that sync with the slides (another way around)

**Comment by cyyynthia** [comment]
Hm, that would make cursor sync less of an issue for the view side, however I'm worried that'd make using certain features of the presenter mode harder to use:
- If support for synced draggable elements is added, the controls should show on the presenter side (and not necessarily on the view side)
- If support for synced Monaco editors is added, the run button should be clickable on the presenter side (and not necessarily on the viewer side)
- All of the above might feel laggy (which can quickly be infuriating when typing code 😅), especially considering #1681 (I've noticed Twitter widgets, when loading, make sync extremely laggy, as a sort-of-consistent lag reproduction) and network-bound deployments as proposed in #1227
- Quickswitching via the overview (when answering questions abt the topic and what not) means bringing back the native cursor to the presenter view
- Making edits to notes requires bringing back the native cursor
- Controlling the timer requires bringing back the native cursor

The cursor is definitely a non-trivial issue (especially if the goal is to support everything needed to make the cursor "feel" like the native cursor), but I think it's best to keep the native cursor on the presenter view, for convenience and ease of use (and lag-free experience, it's less of a concern if the view lags than what the presenter is using imo)

**Comment by antfu** [comment]
I made some changes and released them in `0.51.0-beta.1`

- Syncing controls. Now, it's possible to have the viewer sync back to the presenter, and it's bi-directionally syncing by default.
- In the presenter, a new mode to use screen capture to mirror the viewer's real display back to the presenter. This means that you can directly have your cursor on the viewer screen to do the interactions without the cursor simulations, while you can see them on your presenter view.

Please give it a try and let me know how it works for you.

**Comment by cyyynthia** [comment]
Woah, that was quick! :D 

I played with it a bit, it's cool and was quite lag free (although the quality wasn't as good as "real" rendering of slides); however I'm inclined to stay by my earlier remarks on presenter controls requiring bringing back the mouse to the presenter screen which isn't the best thing ever.

Also, my understanding is that this is only device-local, correct? If I were to have a laptop acting as the view and a tablet acting as a standalone control device, that wouldn't quite work, right?

**Comment by antfu** [comment]
Yeah, indeed, there are always trade-offs. I think we still need to explore for each specific needs. Do you mind to create a separate issue(s) for us to track them? Thanks

Yes, only works on the same device.

**Comment by cyyynthia** [comment]
Sorry for the wait 😅, I opened separate issues to track these better 👍🏻
